Response-Theory/Self-Consistent-Field/beta.py

```diff
@@ -0,0 +1,351 @@
+        tl2 = 2 * np.trace(U[a, ...].dot(G[c, ...]).dot(U[b, ...])[:nocc, :nocc])
+        tl3 = 2 * np.trace(U[c, ...].dot(G[a, ...]).dot(U[b, ...])[:nocc, :nocc])
+        tr1 = np.trace(U[c, ...].dot(U[b, ...]).dot(E[a, ...])[:nocc, :nocc])
+        tr2 = np.trace(U[b, ...].dot(U[c, ...]).dot(E[a, ...])[:nocc, :nocc])
+        tr3 = np.trace(U[c, ...].dot(U[a, ...]).dot(E[b, ...])[:nocc, :nocc])
+        tr4 = np.trace(U[a, ...].dot(U[c, ...]).dot(E[b, ...])[:nocc, :nocc])
+        tr5 = np.trace(U[b, ...].dot(U[a, ...]).dot(E[c, ...])[:nocc, :nocc])
+        tr6 = np.trace(U[a, ...].dot(U[b, ...]).dot(E[c, ...])[:nocc, :nocc])
+        tl = tl1 + tl2 + tl3
+        tr = tr1 + tr2 + tr3 + tr4 + tr5 + tr6
+        hyperpolarizability[r, a] = -2 * (tl - tr)
+
+ref_static = np.array([
+    [ 0.00000001,   0.00000000,   0.22843772],
+    [ 0.00000000,   0.00000000, -25.35476040],
+    [ 0.00000000,   0.00000000, -10.84023375],
+    [ 0.00000000,   0.00000000,   0.00000000],
+    [ 0.22843772,   0.00000000,   0.00000000],
+    [ 0.00000000, -25.35476040,   0.00000000]
+])
+assert np.allclose(ref_static, hyperpolarizability, rtol=0.0, atol=1.0e-3)
+print('\nFirst dipole hyperpolarizability (static):')
+print(hyperpolarizability)
+
+print('Setting up for second-harmonic generation (SHG) calculation...')
+f1 = 0.0773178
+f2 = 2 * f1
+
+print('\nForming response vectors for {} a.u.'.format(f1))
+helper1 = helper_CPHF(mol)
+helper1.solve_dynamic_direct(omega=f1)
+helper1.form_polarizability()
+print(helper1.polar)
+print('\nForming response vectors for {} a.u.'.format(f2))
+helper2 = helper_CPHF(mol)
+helper2.solve_dynamic_direct(omega=f2)
+helper2.form_polarizability()
+print(helper2.polar)
+
+rspvecs1 = helper1.x
+rspvecs2 = helper2.x
+
+# repack response vectors to [norb, norb]
+U1 = np.zeros_like(integrals_mo)
+U2 = np.zeros_like(integrals_mo)
+for i in range(ncomp):
+    U1[i, :nocc, nocc:] = rspvecs1[i][nov:].reshape(nocc, nvir)
+    U1[i, nocc:, :nocc] = rspvecs1[i][:nov].reshape(nocc, nvir).T
+    U2[i, :nocc, nocc:] = rspvecs2[i][nov:].reshape(nocc, nvir)
+    U2[i, nocc:, :nocc] = rspvecs2[i][:nov].reshape(nocc, nvir).T
+
+G1 = np.empty_like(U1)
+G2 = np.empty_like(U2)
+R1_l = psi4.core.Matrix(nbf, nocc)
+R1_r = psi4.core.Matrix(nbf, nocc)
+R2_l = psi4.core.Matrix(nbf, nocc)
+R2_r = psi4.core.Matrix(nbf, nocc)
+npR1_l = np.asarray(R1_l)
+npR1_r = np.asarray(R1_r)
+npR2_l = np.asarray(R2_l)
+npR2_r = np.asarray(R2_r)
+jk.C_clear()
+jk.C_left_add(Co)
+jk.C_right_add(R1_l)
+jk.C_left_add(Co)
+jk.C_right_add(R1_r)
+jk.C_left_add(Co)
+jk.C_right_add(R2_l)
+jk.C_left_add(Co)
+jk.C_right_add(R2_r)
+nCo = np.asarray(Co)
+# Do 4 Fock builds at a time: X/Y vectors for both frequencies; loop
+# over operator components
+for i in range(3):
+    V = integrals_mo[i, ...]
+
+    x1 = U1[i, :nocc, :]
+    y1 = U1[i, :, :nocc]
+    x2 = U2[i, :nocc, :]
+    y2 = U2[i, :, :nocc]
+    npR1_l[...] = C.dot(x1.T)
```

Comment:
I see what you are doing here now, perhaps a better operator would be `npR1_l[:]` instead of Ellipses which typically denote striding.